In [1]:
import pandas as pd 
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [147]:
df=pd.read_csv('PD 2023 Wk 1 Input.csv')

In [148]:
df.head()

Transaction Code  Value  Customer Code  Online or In-Person  \
0  DTB-716-679-576   1448         100001                    2   
1   DS-795-814-303   7839         100001                    2   
2  DSB-807-592-406   5520         100005                    1   
3   DS-367-545-264   7957         100007                    2   
4  DSB-474-374-857   5375         100000                    2   

      Transaction Date  
0  20/03/2023 00:00:00  
1  15/11/2023 00:00:00  
2  14/07/2023 00:00:00  
3  18/08/2023 00:00:00  
4  26/08/2023 00:00:00

In [149]:
df.dtypes

Transaction Code       object
Value                   int64
Customer Code           int64
Online or In-Person     int64
Transaction Date       object
dtype: object

In [150]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365 entries, 0 to 364
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Transaction Code     365 non-null    object
 1   Value                365 non-null    int64 
 2   Customer Code        365 non-null    int64 
 3   Online or In-Person  365 non-null    int64 
 4   Transaction Date     365 non-null    object
dtypes: int64(3), object(2)
memory usage: 14.4+ KB


In [151]:
df.describe()

Value  Customer Code  Online or In-Person
count   365.000000     365.000000           365.000000
mean   4938.813699  100004.961644             1.528767
std    2940.973461       3.277237             0.499857
min      13.000000  100000.000000             1.000000
25%    2321.000000  100002.000000             1.000000
50%    4743.000000  100005.000000             2.000000
75%    7427.000000  100008.000000             2.000000
max    9967.000000  100010.000000             2.000000

In [152]:
df.drop_duplicates().size

1825

In [153]:
df[df.duplicated(keep='first')].size

0

- there are no rows which are completely duplicates 
- there are not null values in the dataset

In [154]:
df['Online or In-Person'].value_counts()

Online or In-Person
2    193
1    172
Name: count, dtype: int64

- online or In person seems like a categorical value indicating if the its a online customer or a In-person customer
- since there are only 2 categories we can keep them as label endcoded itself. (1 - for online and 2 - for In-person as mentioned the requirements)

In [155]:
df['Transaction Code'].value_counts().shape

(365,)

- transaction code seems to be unique for each transaction, and from the requirements we can confirm that the spring of letters in the starting of the transaction code (until the first -) represent the bank with processed that transaction. we can split the transaction code and create a new field called bank. 

In [156]:
df['Transaction Code'].head()

0    DTB-716-679-576
1     DS-795-814-303
2    DSB-807-592-406
3     DS-367-545-264
4    DSB-474-374-857
Name: Transaction Code, dtype: object

In [157]:
def get_bank(x):
    temp=x.split('-')
    if len(temp)>0:
        return temp[0]
    return None

In [158]:
df['Bank']=df['Transaction Code'].apply(get_bank)

In [159]:
df.head()

Transaction Code  Value  Customer Code  Online or In-Person  \
0  DTB-716-679-576   1448         100001                    2   
1   DS-795-814-303   7839         100001                    2   
2  DSB-807-592-406   5520         100005                    1   
3   DS-367-545-264   7957         100007                    2   
4  DSB-474-374-857   5375         100000                    2   

      Transaction Date Bank  
0  20/03/2023 00:00:00  DTB  
1  15/11/2023 00:00:00   DS  
2  14/07/2023 00:00:00  DSB  
3  18/08/2023 00:00:00   DS  
4  26/08/2023 00:00:00  DSB

- bank names have been extracted, looks like categorical values

In [160]:
df['Bank'].value_counts()

Bank
DTB    135
DS     128
DSB    102
Name: count, dtype: int64

In [161]:
df['Bank'].dropna().shape

(365,)

- Bank is categorical and there are no null values, there are 3 banks with DTB bank with 135 transations and DSB bank with 102 transactions. 

In [162]:
df.head()

Transaction Code  Value  Customer Code  Online or In-Person  \
0  DTB-716-679-576   1448         100001                    2   
1   DS-795-814-303   7839         100001                    2   
2  DSB-807-592-406   5520         100005                    1   
3   DS-367-545-264   7957         100007                    2   
4  DSB-474-374-857   5375         100000                    2   

      Transaction Date Bank  
0  20/03/2023 00:00:00  DTB  
1  15/11/2023 00:00:00   DS  
2  14/07/2023 00:00:00  DSB  
3  18/08/2023 00:00:00   DS  
4  26/08/2023 00:00:00  DSB

- transaction date is of type object, we can change to datetime for further manipulations

In [163]:
df['Transaction Date']=pd.to_datetime(df['Transaction Date'])

C:\Users\punith kumar\AppData\Local\Temp\ipykernel_12128\204886444.py:1: UserWarning: Parsing dates in %d/%m/%Y %H:%M:%S format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df['Transaction Date']=pd.to_datetime(df['Transaction Date'])


In [164]:
df.dtypes

Transaction Code               object
Value                           int64
Customer Code                   int64
Online or In-Person             int64
Transaction Date       datetime64[ns]
Bank                           object
dtype: object

- as per requirement, need to create a new field called transaction Date which should contain the day of the transaction

In [165]:
df['Transaction Date']=df['Transaction Date'].dt.day_name()

In [166]:
df.head()

Transaction Code  Value  Customer Code  Online or In-Person  \
0  DTB-716-679-576   1448         100001                    2   
1   DS-795-814-303   7839         100001                    2   
2  DSB-807-592-406   5520         100005                    1   
3   DS-367-545-264   7957         100007                    2   
4  DSB-474-374-857   5375         100000                    2   

  Transaction Date Bank  
0           Monday  DTB  
1        Wednesday   DS  
2           Friday  DSB  
3           Friday   DS  
4         Saturday  DSB

- as per requirements, online or inperson field needs to be changed into categorical field( which appears to be label encoded need to be reverted)

In [167]:
df['Online or In-Person'].dtype

dtype('int64')

In [168]:
df['Online or In-Person']=df['Online or In-Person'].replace([1,2],['Online', 'In-Person'])

In [169]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365 entries, 0 to 364
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Transaction Code     365 non-null    object
 1   Value                365 non-null    int64 
 2   Customer Code        365 non-null    int64 
 3   Online or In-Person  365 non-null    object
 4   Transaction Date     365 non-null    object
 5   Bank                 365 non-null    object
dtypes: int64(2), object(4)
memory usage: 17.2+ KB


In [170]:
df.head()

Transaction Code  Value  Customer Code Online or In-Person Transaction Date  \
0  DTB-716-679-576   1448         100001           In-Person           Monday   
1   DS-795-814-303   7839         100001           In-Person        Wednesday   
2  DSB-807-592-406   5520         100005              Online           Friday   
3   DS-367-545-264   7957         100007           In-Person           Friday   
4  DSB-474-374-857   5375         100000           In-Person         Saturday   

  Bank  
0  DTB  
1   DS  
2  DSB  
3   DS  
4  DSB

In [184]:
df_group_by_bank=df[['Bank', 'Value']].groupby(by='Bank').sum().sort_values(by='Value').reset_index().rename(columns={'Value': 'Total Transactions'})

In [185]:
df_group_by_bank.head()

Bank  Total Transactions
0  DSB              530489
1  DTB              618238
2   DS              653940

In [199]:
df_group2_by_bank=df.groupby(by=['Bank','Online or In-Person', 'Transaction Date'])['Value'].sum().reset_index()

In [202]:
print(df_group2_by_bank.shape)
df_group2_by_bank.head()

(42, 4)


Bank Online or In-Person Transaction Date  Value
0   DS           In-Person           Friday  58599
1   DS           In-Person           Monday  42806
2   DS           In-Person         Saturday  34867
3   DS           In-Person           Sunday  51301
4   DS           In-Person         Thursday  75582

In [204]:
df_group3_by_bank=df[['Bank', 'Customer Code', 'Value']].groupby(by=['Bank', 'Customer Code']).sum().reset_index()

In [206]:
print(df_group3_by_bank.shape)
df_group3_by_bank.head()

(33, 3)


Bank  Customer Code  Value
0   DS         100000  57909
1   DS         100001  53063
2   DS         100002  69803
3   DS         100003  25482
4   DS         100004  63315